### PR TITLE
Change the fetch method to deal with recyclable key cache strategy

### PR DESCRIPTION
**Pull request by cintamani**
In order to keep compatibility between the AMS cache feature and with Rails > 5.1  cache versioning which introduces Recyclable Keys, we have to account for the version of the cache separately from the cache key.
This PR address the issue for application with Rails version > 5.1 where the `cache_versioning` setting is `true`

More info: https://github.com/rails-api/active_model_serializers/issues/2287

### CODE REVIEW

**Comment by bf4** [comment]
Would you mind rebasing off of 0-10-stable?

**Comment by cintamani** [comment]
@bf4 Hi there :) this is already rebased against 0-10-stable

**Comment by bf4** [comment]
Looking at AppVeyor failures on Ruby 2.3

```
Finished in 2.835022s, 199.9985 runs/s, 345.6764 assertions/s.
  1) Failure:
ActiveModelSerializers::CacheTest#test_expiring_of_cache_at_update_of_record [C:/projects/active-model-serializers/test/cache_test.rb:165]:
--- expected
+++ actual
@@ -1,2 +1,2 @@
 # encoding: UTF-8
-"Bar"
+"Foo"
```

**Comment by cintamani** [comment]
@bf4 can I ask some more details on appveyor? I'm trying to reproduce the issue running the tests in 2.3 ruby and 4.2 rails but without success. I have added some more code based on what I have found here: https://github.com/rails-api/active_model_serializers/pull/1168

File: test/cache_test.rb [hunk 1]
```diff
@@ -144,6 +149,28 @@ class InheritedRoleSerializer < RoleSerializer
...
+
+      author.update_attributes(name: 'Bar')
+      author_json = AuthorSerializerWithCache.new(author).as_json
+
+      expected = 'Bar'
+      actual = author_json[:name]
+      if ENV['APPVEYOR'] && actual != expected
+        skip('Cache expiration tests sometimes fail on Appveyor. FIXME :)')
+      else
+        assert_equal actual, expected
```
[the remark this answers is not in the document]
Comment (cintamani) [reply test/cache_test.rb]: Thanks! Sorry I am an rspec user 🙈  😂

[the remark this answers is not in the document]
Comment (cintamani) [reply test/cache_test.rb]:
```suggestion
        assert_equal expected, actual
```

File: lib/active_model/serializer/concerns/caching.rb [hunk 2]
```diff
@@ -229,6 +229,7 @@ def fetch_attributes(fields, cached_attributes, adapter_instance)
       def fetch(adapter_instance, cache_options = serializer_class._cache_options, key = nil)
         if serializer_class.cache_store
           key ||= cache_key(adapter_instance)
+          cache_options = (cache_options || {}).merge(version: object_cache_version) if object_cache_version
```
[the remark this answers is not in the document]
Comment (cintamani) [reply lib/active_model/serializer/concerns/caching.rb]: Do you think it is likely to create issues in other parts of the code?

[the remark this answers is not in the document]
Comment (bf4) [reply lib/active_model/serializer/concerns/caching.rb]: I'm not familiar enough with the Rails code changes to be really sure.

I'd look at https://blog.heroku.com/cache-invalidation-rails-5-2-dalli-store to get any ideas.

How does it fail without this PR?

[the remark this answers is not in the document]
Comment (bf4) [reply lib/active_model/serializer/concerns/caching.rb]: @cintamani ?

[the remark this answers is not in the document]
Comment (cintamani) [reply lib/active_model/serializer/concerns/caching.rb]: @bf4 I have added a test that you can see failing without the rest of the fixes in this pr: https://github.com/rails-api/active_model_serializers/pull/2288/files#diff-337d3920ba1fab97fd9fdd688d7664f3R152  :)

File: test/cache_test.rb [hunk 3]
```diff
@@ -144,6 +149,28 @@ class InheritedRoleSerializer < RoleSerializer
...
+      if ARModels::Author.respond_to?(:cache_versioning)
+        ARModels::Author.cache_versioning = true
+      end
+
+      author = ARModels::Author.create(name: 'Foo')
+      author_json = AuthorSerializerWithCache.new(author).as_json
+
+      assert_equal 'Foo', author_json[:name]
+
+      author.update_attributes(name: 'Bar')
```
Comment (bf4) [inline test/cache_test.rb]: Is the reason for the db operation here to break the cache?

btw, should be `update` not `update_attributes`, and better to do `update!` to abort the test if a validation fails (not that it is likely, but still..)

File: test/cache_test.rb [hunk 4]
```diff
@@ -144,6 +149,28 @@ class InheritedRoleSerializer < RoleSerializer
       @blog_serializer     = BlogSerializer.new(@blog)
     end
 
+    def test_expiring_of_cache_at_update_of_record
+      if ARModels::Author.respond_to?(:cache_versioning)
+        ARModels::Author.cache_versioning = true
+      end
```
Comment (bf4) [inline test/cache_test.rb]: This fails without the cache versioning? why?

File: test/cache_test.rb (hunk 3 above)
Comment (cintamani) [reply test/cache_test.rb]: `Is the reason for the db operation here to break the cache?` yes updating an attribute should invalidate the cache :) this is what's broken currently for rails apps that have `cache_versioning` setted to `true`

File: test/cache_test.rb (hunk 4 above)
Comment (cintamani) [reply test/cache_test.rb]: 😅 I think I specified it in the description :P 

It fails because the current cache implementation rely on the `cache_key` provided by ActiveRecord to contains the `timestamp` but when `cache_versioning` is true the key of course does not contain the `timestamp`. 

`cache_versioning` changes the strategy of caching for rails implementing "reusable keys". This means that the key never change, and the content of the cache instead contains the version (timestamp in this case).

**Comment by bf4** [comment]
Ok, I'm happy to merge this when you're happy with it

**Comment by cintamani** [comment]
@bf4 thanks 🙇‍♀️ ASAP I have some time to test it one last time I'll merge it :)

**Comment by bf4** [comment]
@cintamani would you like this merged now? (I'm not confident I understand your last comment as intended)

File: lib/active_model/serializer/concerns/caching.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (bf4) [reply lib/active_model/serializer/concerns/caching.rb]: > do we really need to send the version option here ?

@wasifhossain are you saying you think this PR takes the wrong approach?

It's unclear to me if your review is based on usage of recyclable cache keys or not. I believe @cintamani is using this in production

**Comment by cintamani** [comment]
@bf4 sorry I'm not very active here 🙇‍♀️I have tested it again in my local environment and it still looks good and working as intended. But consider we are not currently using this in production yet due to issues with the Dalli gem. If you are still up for it, I will merge this one in tomorrow.

**Comment by bf4** [comment]
@cintamani Since the pace of development here is pretty slow, I'd prefer waiting for you to confirm it works for you in production before I merging this PR. That ok?

**Comment by cintamani** [comment]
@bf4 totally :) I'll keep you posted

**Comment by maxrosecollins** [comment]
Can we get this merged? 

None of my cached objects ever update even with `belongs_to :parent, touch: true`

**Comment by bf4** [comment]
@maxrosecollins have you confirmed it works for you in prod?

Obviously we can merge it. But I would like confirmation not only that tests pass, but that it solves the problem it's intended to. I'm not currently using AMS.

**Comment by maxrosecollins** [comment]
I tried to 

`gem 'active_model_serializers', github: "cintamani/active_model_serializers", branch: "patch-1"` 

I got this error

```
Fetching https://github.com/cintamani/active_model_serializers.git
fatal: Could not parse object '7427d893f4e07e45003e92e8303bf698764f68fc'.
Revision 7427d893f4e07e45003e92e8303bf698764f68fc does not exist in the repository https://github.com/cintamani/active_model_serializers.git. Maybe you misspelled it?
```

**Comment by bf4** [comment]
@maxrosecollins This isn't really the place to debug bundling a Gem from GitHub...

I see https://github.com/cintamani/active_model_serializers/tree/patch-1 is at 7427d893f4e07e45003e92e8303bf698764f68fc  so I'm not really sure what oddities in your Gemfile are causing the problem.  Try specified the absolute git url, and it that doesn't work, probably try a forum or slack.  `git: "https://github.com/cintamani/active_model_serializers.git"`

**Comment by bf4** [comment]
@maxrosecollins like I said, this isn't the place to debug installing a gem from github. It makes the conversation on the topic of the PR hard to follow.

Also, though I'm not using AMS, I just added the prescribed line to my Gemfile and successfully installed it.  So, whatever is happening on your computer is 99.99999% certain to be unrelated to AMS.

I wish I had the bandwidth to help you further and a place for that, but I don't right now.  I am sorry.

I'm just one person taking time away from his work to try and help make sure changes like this PR help people (and not cause harm)

here my line in my Gemfile. I'm using Ruby 2.4.4, bundler 1.16.1, Rails 5.0.7

```ruby
gem "active_model_serializers", git: "https://github.com/cintamani/active_model_serializers.git", branch: "patch-1"
```

If you found a bug in bundler, they'll ask you to make a test repo to demonstrate and you can post an issue there.

**Comment by maxrosecollins** [comment]
No worries.

I'll take a look and see if I can get it working.

**Comment by cintamani** [comment]
@maxrosecollins this version works for sure https://github.com/rails-api/active_model_serializers/pull/2288/commits/7d498d279e929fc031cdb78148f13b2a27dc04f6 let me know if you get the time to make this PR's code work. I am very full those days and last time I have checked I found the error you are mentioning too. Still didn't have time to look at it properly.

**Comment by wasifhossain** [comment]
> @maxrosecollins this version works for sure [7d498d2](https://github.com/rails-api/active_model_serializers/commit/7d498d279e929fc031cdb78148f13b2a27dc04f6) let me know if you get the time to make this PR's code work. I am very full those days and last time I have checked I found the error you are mentioning too. Still didn't have time to look at it properly.

I can confirm the fact that this commit [7d498d2](https://github.com/rails-api/active_model_serializers/commit/7d498d279e929fc031cdb78148f13b2a27dc04f6) (1st commit of this PR) should work :tada: . Working excellent :ok_hand: on our production app with this cherry-pick https://github.com/gomalindo/active_model_serializers/commit/823b55865710ac6b01b2f85883e0ec31311630a2, i.e. **cache key expires properly** in both cases when `ActiveRecord::Base.cache_versioning` is set to `true/false` as the cache key is formed containing the dynamic part (`updated_at`).

Unfortunately, unlike that single commit, **this PR does not work as expected** :( 

### TL;DR [collection_serializer](https://github.com/rails-api/active_model_serializers/blob/0-10-stable/lib/active_model/serializer/collection_serializer.rb) cache keys don't expire at all using this PR, resulting in **stale cache values**

For those who are not concerned yet what this PR is trying to solve, I would humbly request to have a quick visit to these links
* https://github.com/rails/rails/pull/29092: PR which introduced `recyclable cache keys` in **rails 5.2**
* https://blog.heroku.com/cache-invalidation-rails-5-2-dalli-store: a recent issue (does not relate to this PR directly, but will help in debugging similar issues)

In short, when `ActiveRecord::Base.cache_versioning = true` (which will become default from **Rails 6**), `updated_at` is stored inside cache value, while cache key remains static over time (which is why its called `recyclable cache key`), e.g. `products/1`. In AMS, a cache key would look like: `products/1/attributes`. 

But when `ActiveRecord::Base.cache_versioning = false`, cache keys would look like `products/1-20190110212435047688/attributes`

To read/write the latest cache value, we need to provide the `:version` option in the [ActiveSupport::Cache::Store#fetch](https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-fetch) method, which can be easily retrieved by calling [ActiveRecord::Base#cache_version](https://api.rubyonrails.org/classes/ActiveRecord/Integration.html#method-i-cache_version)

Here are some examples to highlight the behavior of a recyclable cache key
```
# returns up-to-date cached value
def cached_product(product)
  Rails.cache.fetch(product.cache_key, version: product.cache_version) do
    product
  end
end

product = Product.first
product.name # => "Product 1"
v1 = product.cache_version # => "20190125020615012258"

# here calling this method will store the current product value inside `product.cache_key` ('products/1')
cached_product(product).name # => "Product 1"

# now reading the cache value with/without the cache version will return v1 value
Rails.cache.read(product.cache_key, version: v1).name # => "Product 1"
Rails.cache.read(product.cache_key).name # => "Product 1"

# let's update the product
product.update(name: 'Product 2')
product.name # => "Product 2"
v2 = product.cache_version # => "20190125022443338534"

# here calling this method will overwrite the cache value with the v2 product value
cached_product(product).name # => "Product 2"

# now reading the cache value with/without the cache version will return v2 value
Rails.cache.read(product.cache_key, version: v2).name # => "Product 2"
Rails.cache.read(product.cache_key).name # => "Product 2"

# but trying to read the cache value with `version: v1` will return nil as it's been overwritten with v2
# hence ensuring that the cache key is recycled properly
Rails.cache.read(product.cache_key, version: v1) # => nil
Rails.cache.read(product.cache_key).name # => "Product 2"
```

This is what the PR tried to do exactly: https://github.com/rails-api/active_model_serializers/pull/2288/files#diff-b762caa6dd51036307106006dd81600f. So the version info is passed while fetching the cache value.

Now we can discuss the actual issue. For collection serializer, the serialization happens in this sequence:
1. all the attributes' keys & values are [read from the cache](https://github.com/rails-api/active_model_serializers/blob/7427d893f4e07e45003e92e8303bf698764f68fc/lib/active_model/serializer/concerns/caching.rb#L174) using [ActiveSupport::Cache::Store#read_multi](https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-read_multi)
```
# active_model_serializers/lib/active_model/serializer/collection_serializer.rb#25

options[:cached_attributes] ||= ActiveModel::Serializer.cache_read_multi(self, adapter_instance, options[:include_directive])
```
2. [here](https://github.com/rails-api/active_model_serializers/blob/7427d893f4e07e45003e92e8303bf698764f68fc/lib/active_model/serializer/concerns/caching.rb#L222) it looks up a key in the hash derived above. If the key is found, it returns the value. Otherwise the new value is written to cache using the cache version introduced in this PR.

Here the thing to note is: once the cache is written for some key, it will be read using `read_multi` as described on point 1. But **`ActiveSupport::Cache::Store#read_multi` unfortunately does not support versioning yet.** (does not accept `:version` option unlike `fetch/read`). And so we will always get the **oldest cache value** stored using the recyclable cache key support introduced in this PR.

Summary (assuming a model has been updated multiple times):
* Without this PR: 
  * `ActiveRecord::Base.cache_versioning = true`: cache key looks like `products/1/attributes` resulting in **stale cache value**
  * `ActiveRecord::Base.cache_versioning = false`: cache key looks like `products/1-20190110212435047688/attributes` resulting in **up-to-date cache value**
* With this PR:
  * `ActiveRecord::Base.cache_versioning = true`: cache key looks like `products/1/attributes` resulting in **stale cache value**
  * `ActiveRecord::Base.cache_versioning = false`: cache key looks like `products/1/attributes` resulting in **stale cache value**
* With the very first commit in this PR [7d498d2](https://github.com/rails-api/active_model_serializers/commit/7d498d279e929fc031cdb78148f13b2a27dc04f6):
  * `ActiveRecord::Base.cache_versioning = true`: cache key looks like `products/1-20190110212435047688/attributes` resulting in **up-to-date cache value**
  * `ActiveRecord::Base.cache_versioning = false`: cache key looks like `products/1-20190110212435047688/attributes` resulting in **up-to-date cache value**

So my humble opinion is that we stick with the form of cache key that would embed the version info itself as before i.e `products/1-20190110212435047688/attributes` which is fixed with the very first commit in this PR [7d498d2](https://github.com/rails-api/active_model_serializers/commit/7d498d279e929fc031cdb78148f13b2a27dc04f6) and wait for the rails community until they support recyclable cache key versioning in [ActiveSupport::Cache::Store#read_multi](https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-read_multi)

**Comment by cintamani** [comment]
Yes. The idea behind is that we start sending the key and versions as 2 separate value to the cache gem. But since not even `dalli` is actually ready to deal with recyclable keys, I believe we should merge the first version of this change that simply uses `cache_key_with_version` when available. 
If you all agree, I'll update the code.

**Comment by cintamani** [comment]
And most important, @wasifhossain thanks a lot for taking the time to write down all that. Really helpful :)

**Comment by bf4** [comment]
@wasifhossain @cintamani Awesome job on this!

@cintamani if you're too busy to get the working code into a mergeable state, would you like me to ask @wasifhossain if they'd like to make a PR to replace this one?  I'd assume @wasifhossain would just pull in your commits to make sure you get credit in the commit history. Or @wasifhossain could make a PR into your fork directly, but they that's strictly between you two until it is merged.

**Comment by cintamani** [comment]
@bf4 I will do it today. I will have to just put the commits back to the first working version :)

**Comment by cintamani** [comment]
@bf4 do you have any idea of what's wrong with travis?

**Comment by bf4** [comment]
@cintamani I merged in a fix for CI https://github.com/rails-api/active_model_serializers/pull/2310  this ready to be merged? @wasifhossain you're happy with this?

**Comment by cintamani** [comment]
@bf4 cool let me rebase and let's see if Travis is happy this is ready to merge

**Comment by wasifhossain** [comment]
@bf4 now the code seems to work properly for all the cases related to caching. In order to avoid any regression caused by future changes related to caching, I would highly appreciate if both edge cases (that I tried to explain in my comment above) are covered in tests too

* `ActiveRecord::Base.cache_versioning = true`: cache key looks like `products/1-20190110212435047688/attributes` resulting in **up-to-date cache value**
* `ActiveRecord::Base.cache_versioning = false`: cache key looks like `products/1-20190110212435047688/attributes` resulting in **up-to-date cache value**

Most importantly, the test should try to **update a record multiple times** and see if caching still works properly.

@cintamani I am eager to know your thoughts on this.

Thanks

**Comment by cintamani** [comment]
@wasifhossain take a look at my changes to the tests in this Pull Request. I believe all the scenarios above are covered by new and existing tests.

**Comment by wasifhossain** [comment]
@cintamani I agree with you that existing tests are good enough to check if cache expires properly for a **single record**. Unfortunately those tests fail to cover the **collection_serializer** case I detailed in my comment [above](https://github.com/rails-api/active_model_serializers/pull/2288#issuecomment-457445111).

So I am going to add a tiny commit containing the missing piece. With its addition, I believe that we are good enough to avoid any regression caused by future changes related to caching.

Please note that, if you run this new test against the temporary changes introduced after the 1st commit in this PR (reverted at this point), you will find the test would fail.

Will look forward to hearing your feedback. Feel free to modify the test as you find it appropriate.

Thanks

**Comment by cintamani** [comment]
@wasifhossain oh now I get it! :)
Please feel free to push your tests to commit against this change. It looks good to me.

**Comment by wasifhossain** [comment]
thats really good to know :raised_hands: @cintamani 
I think I don't have access to push commits to your branch. Feel free to include the changes in your own commit if you like :)

**Comment by bf4** [comment]
@wasifhossain fwiw you can fork @cintamani repo and make a pr into the branch @cintamani is making this PR from (patch-1)

**Comment by wasifhossain** [comment]
oh I see :smile: thanks for pointing @bf4 . I will do that by tonight.

**Comment by wasifhossain** [comment]
@cintamani here is the PR: https://github.com/cintamani/active_model_serializers/pull/1

**Comment by bf4** [comment]
@cintamani @wasifhossain if you're both happy with this, I can merge it and make a release.

Would either of you like to be added as collaborators? You've done good work here

File: test/cache_test.rb [hunk 5]
```diff
@@ -146,6 +151,49 @@ class InheritedRoleSerializer < RoleSerializer
...
+      foo2              = 'Foo2'
+      author            = ARModels::Author.create(name: foo)
+      author2           = ARModels::Author.create(name: foo2)
+      author_collection = [author, author, author2]
+
+      collection_json = render_object_with_cache(author_collection, each_serializer: AuthorSerializerWithCache)
+      assert_equal [{ name: foo }, { name: foo }, { name: foo2 }], collection_json
+
+      bar = 'Bar'
+      author.update_attributes(name: bar)
```
[the remark this answers is not in the document]
Comment (cintamani) [reply test/cache_test.rb]: Oh I'm sorry! I have missed that! Let's try this:

```suggestion
      author.update!(name: bar)
```

File: test/cache_test.rb [hunk 6]
```diff
@@ -146,6 +151,49 @@ class InheritedRoleSerializer < RoleSerializer
...
+      if ARModels::Author.respond_to?(:cache_versioning)
+        ARModels::Author.cache_versioning = true
+      end
+
+      author = ARModels::Author.create(name: 'Foo')
+      author_json = AuthorSerializerWithCache.new(author).as_json
+
+      assert_equal 'Foo', author_json[:name]
+
+      author.update_attributes(name: 'Bar')
```
Comment (bf4) [inline test/cache_test.rb]:
```suggestion
      author.update!(name: 'Bar')
```

**Comment by cintamani** [comment]
If Trevis is happy, I'm happy for you to merge this. And yes I would like to be added as a collaborator if you are happy with it :)

File: test/cache_test.rb (hunk 6 above)
Comment (cintamani) [reply test/cache_test.rb]: Trying to understand why the change to the first `update!` is failing on Travis first

**Comment by cintamani** [comment]
@bf4 looks like we have lost support to 4.2 (also is Appveyor suppose to run Rails 4.2 only 🤔?)

**Comment by wasifhossain** [comment]
@bf4 i would also be very happy to be a collaborator. Would appreciate some directions from you to be effective enough with the role. Thanks

**Comment by bf4** [comment]
I've sent you both invites

**Comment by cintamani** [comment]
@bf4 Rails 4.2 versions are in the list of "Allowed Failures" for travis so I suggest we update the rails version on AppVeyor too. Also the 4.2 rails version failures doesn't look related to the code in this PR, can you confirm it? Also, can you confirm that we are happy for the next released version of AMS to not support Rails 4.2?

File: test/cache_test.rb [hunk 7]
```diff
@@ -146,6 +151,49 @@ class InheritedRoleSerializer < RoleSerializer
       @blog_serializer     = BlogSerializer.new(@blog)
     end
 
+    def test_expiring_of_cache_at_update_of_record
+      if ARModels::Author.respond_to?(:cache_versioning)
+        ARModels::Author.cache_versioning = true
```
Comment (bf4) [inline test/cache_test.rb]: we probably want to just change this value in the context of the test.  

```ruby
    def test_expiring_of_cache_at_update_of_record
      if ARModels::Author.respond_to?(:cache_versioning)
        original_cache_versioning = ARModels::Author.cache_versioning
        ARModels::Author.cache_versioning = true
     else
       original_cache_versioning = :no_cache_versioning
     end
     .....
   ensure
     ARModels::Author.cache_versioning = original_cache_versioning unless original_cache_versioning == :no_cache_versioning
   end
```

**Comment by bf4** [comment]
@cintamani 
> Rails 4.2 versions are in the list of "Allowed Failures" for travis

? https://github.com/rails-api/active_model_serializers/blob/209834dc49d4865bf9797c51f086a5f05718cb44/.travis.yml#L47-L62

File: test/cache_test.rb (hunk 7 above)
Comment (cintamani) [reply test/cache_test.rb]: neat!

**Comment by cintamani** [comment]
@bf4 sorry my mistake. The way Travis show the `allowed failures` in their view make me thought it was allowed for the Rails version.

File: test/cache_test.rb (hunk 6 above)
Comment (cintamani) [reply test/cache_test.rb]: I have missed this comment sorry or I would have made the change myself :(

**Comment by bf4** [comment]
wtf failures

```
/home/travis/build/rails-api/active_model_serializers/vendor/bundle/ruby/2.2.0/gems/activerecord-4.1.16/lib/active_record/connection_adapters/connection_specification.rb:190:in `rescue in spec': Specified 'sqlite3' for database adapter, but the gem is not loaded. Add `gem 'sqlite3'` to your Gemfile (and ensure its version is at the minimum required by ActiveRecord). (Gem::LoadError)
```

**Comment by cintamani** [comment]
Yep saw it on Appveyor first but I ignored it. I see Travis is having the same issue 😩investigating

**Comment by bf4** [comment]
@cintamani the good news it, you now have the power to create a PR to fix it and merge when green :)

File: CHANGELOG.md [hunk 8]
```diff
@@ -7,6 +7,7 @@ Breaking changes:
 Features:
 
 Fixes:
+  - [#2288](https://github.com/rails-api/active_model_serializers/pull/2288). Fixes #2287. (@cintamani)
```
Comment (cintamani) [inline CHANGELOG.md]:
```suggestion
  - [#2288](https://github.com/rails-api/active_model_serializers/pull/2288). Fixes #2287. (@cintamani) 
```

**Comment by cintamani** [comment]
We are very unlucky! Looks like the issue is the newly released version of sqlite3 we are using in our test. It was released on the 4th of February, just in time to mess up with this PR 😆 
I have added a version rule in the Gemfile so we stick to 1.3.13. Appveyor and Travis looks happy again 🎊 🎊

**Comment by cintamani** [comment]
@bf4 will you deal with creating a new release? Otherwise I'll give it a go myself

**Comment by bf4** [comment]
@cintamani you can't release until you get rubygems permissions.  I'm going to wait a little bit before adding you as a gem owner.

I juste released 0.10.9

https://github.com/rails-api/active_model_serializers/commit/bb0f9d088df242a65d4e68c6f309875e589f0638

https://twitter.com/rubygems/status/1093924332659789827

cc @wasifhossain

**Comment by cintamani** [comment]
Thank you, @bf4! I have never dealt with maintaining a public gem before, so forgive my ignorance about the RubyGems permissions 🙂

**Comment by bf4** [comment]
@cintamani it's a pretty common thing to forget.

The gems we install are from rubygems.org. Only gem owners can 'push' a gem
GitHub is just where a lot of people collaborate on code. But permissions to write code on GitHub don't sync in any way with ownership on rubygems.org

**Comment by shaojunda** [comment]
> > @maxrosecollins this version works for sure [7d498d2](https://github.com/rails-api/active_model_serializers/commit/7d498d279e929fc031cdb78148f13b2a27dc04f6) let me know if you get the time to make this PR's code work. I am very full those days and last time I have checked I found the error you are mentioning too. Still didn't have time to look at it properly.
> 
> I can confirm the fact that this commit [7d498d2](https://github.com/rails-api/active_model_serializers/commit/7d498d279e929fc031cdb78148f13b2a27dc04f6) (1st commit of this PR) should work 🎉 . Working excellent 👌 on our production app with this cherry-pick [gomalindo@823b558](https://github.com/gomalindo/active_model_serializers/commit/823b55865710ac6b01b2f85883e0ec31311630a2), i.e. **cache key expires properly** in both cases when `ActiveRecord::Base.cache_versioning` is set to `true/false` as the cache key is formed containing the dynamic part (`updated_at`).
> 
> Unfortunately, unlike that single commit, **this PR does not work as expected** :(
> 
> ### TL;DR [collection_serializer](https://github.com/rails-api/active_model_serializers/blob/0-10-stable/lib/active_model/serializer/collection_serializer.rb) cache keys don't expire at all using this PR, resulting in **stale cache values**
> For those who are not concerned yet what this PR is trying to solve, I would humbly request to have a quick visit to these links
> 
> * [rails/rails#29092](https://github.com/rails/rails/pull/29092): PR which introduced `recyclable cache keys` in **rails 5.2**
> * https://blog.heroku.com/cache-invalidation-rails-5-2-dalli-store: a recent issue (does not relate to this PR directly, but will help in debugging similar issues)
> 
> In short, when `ActiveRecord::Base.cache_versioning = true` (which will become default from **Rails 6**), `updated_at` is stored inside cache value, while cache key remains static over time (which is why its called `recyclable cache key`), e.g. `products/1`. In AMS, a cache key would look like: `products/1/attributes`.
> 
> But when `ActiveRecord::Base.cache_versioning = false`, AMS cache keys would look like `products/1-20190110212435047688/attributes`
> 
> To read/write the latest cache value, we need to provide the `:version` option in the [ActiveSupport::Cache::Store#fetch](https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-fetch) method, which can be easily retrieved by calling [ActiveRecord::Base#cache_version](https://api.rubyonrails.org/classes/ActiveRecord/Integration.html#method-i-cache_version)
> 
> Here are some examples to highlight the behavior of a recyclable cache key
> 
> ```
> # returns up-to-date cached value
> def cached_product(product)
>   Rails.cache.fetch(product.cache_key, version: product.cache_version) do
>     product
>   end
> end
> 
> product = Product.first
> product.name # => "Product 1"
> v1 = product.cache_version # => "20190125020615012258"
> 
> # here calling this method will store the current product value inside `product.cache_key` ('products/1')
> cached_product(product).name # => "Product 1"
> 
> # now reading the cache value with/without the cache version will return v1 value
> Rails.cache.read(product.cache_key, version: v1).name # => "Product 1"
> Rails.cache.read(product.cache_key).name # => "Product 1"
> 
> # let's update the product
> product.update(name: 'Product 2')
> product.name # => "Product 2"
> v2 = product.cache_version # => "20190125022443338534"
> 
> # here calling this method will overwrite the cache value with the v2 product value
> cached_product(product).name # => "Product 2"
> 
> # now reading the cache value with/without the cache version will return v2 value
> Rails.cache.read(product.cache_key, version: v2).name # => "Product 2"
> Rails.cache.read(product.cache_key).name # => "Product 2"
> 
> # but trying to read the cache value with `version: v1` will return nil as it's been overwritten with v2
> # hence ensuring that the cache key is recycled properly
> Rails.cache.read(product.cache_key, version: v1) # => nil
> Rails.cache.read(product.cache_key).name # => "Product 2"
> ```
> 
> This is what the PR tried to do exactly: https://github.com/rails-api/active_model_serializers/pull/2288/files#diff-b762caa6dd51036307106006dd81600f. So the version info is passed while fetching the cache value.
> 
> Now we can discuss the actual issue. For collection serializer, the serialization happens in this sequence:
> 
> 1. all the attributes' keys & values are [read from the cache](https://github.com/rails-api/active_model_serializers/blob/7427d893f4e07e45003e92e8303bf698764f68fc/lib/active_model/serializer/concerns/caching.rb#L174) using [ActiveSupport::Cache::Store#read_multi](https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-read_multi)
> 
> ```
> # active_model_serializers/lib/active_model/serializer/collection_serializer.rb#25
> 
> options[:cached_attributes] ||= ActiveModel::Serializer.cache_read_multi(self, adapter_instance, options[:include_directive])
> ```
> 
> 1. [here](https://github.com/rails-api/active_model_serializers/blob/7427d893f4e07e45003e92e8303bf698764f68fc/lib/active_model/serializer/concerns/caching.rb#L222) it looks up a key in the hash derived above. If the key is found, it returns the value. Otherwise the new value is written to cache using the cache version introduced in this PR.
> 
> Here the thing to note is: once the cache is written for some key, it will be read using `read_multi` as described on point 1. But **`ActiveSupport::Cache::Store#read_multi` unfortunately does not support versioning yet.** (does not accept `:version` option unlike `fetch/read`). And so we will always get the **oldest cache value** stored using the recyclable cache key support introduced in this PR.
> 
> Summary (assuming a model has been updated multiple times):
> 
> * Without this PR:
>   
>   * `ActiveRecord::Base.cache_versioning = true`: cache key looks like `products/1/attributes` resulting in **stale cache value**
>   * `ActiveRecord::Base.cache_versioning = false`: cache key looks like `products/1-20190110212435047688/attributes` resulting in **up-to-date cache value**
> * With this PR:
>   
>   * `ActiveRecord::Base.cache_versioning = true`: cache key looks like `products/1/attributes` resulting in **stale cache value**
>   * `ActiveRecord::Base.cache_versioning = false`: cache key looks like `products/1-20190110212435047688/attributes` resulting in **up-to-date cache value**
> * With the very first commit in this PR [7d498d2](https://github.com/rails-api/active_model_serializers/commit/7d498d279e929fc031cdb78148f13b2a27dc04f6):
>   
>   * `ActiveRecord::Base.cache_versioning = true`: cache key looks like `products/1-20190110212435047688/attributes` resulting in **up-to-date cache value**
>   * `ActiveRecord::Base.cache_versioning = false`: cache key looks like `products/1-20190110212435047688/attributes` resulting in **up-to-date cache value**
> 
> So my humble opinion is that we stick with the form of cache key that would embed the version info itself as before i.e `products/1-20190110212435047688/attributes` which is fixed with the very first commit in this PR [7d498d2](https://github.com/rails-api/active_model_serializers/commit/7d498d279e929fc031cdb78148f13b2a27dc04f6) and wait for the rails community until they support recyclable cache key versioning in [ActiveSupport::Cache::Store#read_multi](https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-read_multi)

```
def cached_product(product)
  Rails.cache.fetch(product.cache_key, version: product.cache_version) do
    product
  end
end
```

I have one question, if you already get product, why use cache? @wasifhossain

**Comment by wasifhossain** [comment]
@shaojunda the example was intended to demonstrate the behavior of a **recyclable cache key**.

**Comment by shaojunda** [comment]
I have always wanted to try the recyclable cache key, but I have never figured out how to use it. Do you have an example for use in a production environment?  @wasifhossain
